File: src/satquery/change.ts
import { BoundingBox, MasksMap, RasterScene } from '../types';
import { computeNdvi, computeNdwi, computeNdbi } from './indices';
import { extractBoundingBoxes, morphClean } from './segmentation';

export interface SemanticChangeResult {
  answer: string;
  masks: MasksMap;
  boxes: BoundingBox[];
  dominant: string;
  deltas: {
    ndvi_mean: number;
    ndwi_mean: number;
    ndbi_mean: number;
  };
  method: string;
  confidence: number;
}

export function computeSemanticChange(
  t1: RasterScene,
  t2: RasterScene
): SemanticChangeResult {
  const width = Math.min(t1.width, t2.width);
  const height = Math.min(t1.height, t2.height);
  const len = width * height;

  const ndvi1 = computeNdvi(t1.bands.B08, t1.bands.B04);
  const ndvi2 = computeNdvi(t2.bands.B08, t2.bands.B04);
  const ndwi1 = computeNdwi(t1.bands.B03, t1.bands.B08);
  const ndwi2 = computeNdwi(t2.bands.B03, t2.bands.B08);
  const ndbi1 = computeNdbi(t1.bands.B11, t1.bands.B08);
  const ndbi2 = computeNdbi(t2.bands.B11, t2.bands.B08);

  const rawFlooding = new Uint8Array(len);
  const rawDeforestation = new Uint8Array(len);
  const rawConstruction = new Uint8Array(len);
  const rawCrop = new Uint8Array(len);

  let sumDNdvi = 0;
  let sumDNdwi = 0;
  let sumDNdbi = 0;

  for (let i = 0; i < len; i++) {
    const dNdvi = ndvi2[i] - ndvi1[i];
    const dNdwi = ndwi2[i] - ndwi1[i];
    const dNdbi = ndbi2[i] - ndbi1[i];

    sumDNdvi += dNdvi;
    sumDNdwi += dNdwi;
    sumDNdbi += dNdbi;

    const isFlooding = ndwi2[i] > 0.05 && dNdwi > 0.12 && ndvi2[i] < ndvi1[i];
    const isDeforest = dNdvi < -0.15 && ndvi1[i] > 0.25 && ndwi2[i] < 0.1;
    const isConstruct = dNdbi > 0.12 && dNdvi < -0.05 && ndwi2[i] < 0.1;
    const isCrop =
      Math.abs(dNdvi) > 0.12 &&
      ndvi1[i] > 0.15 &&
      ndvi2[i] > 0.05 &&
      ndwi2[i] < 0.15 &&
      !isConstruct &&
      !isDeforest;

    rawFlooding[i] = isFlooding ? 1 : 0;
    rawDeforestation[i] = isDeforest ? 1 : 0;
    rawConstruction[i] = isConstruct ? 1 : 0;
    rawCrop[i] = isCrop ? 1 : 0;
  }

  const cleanFlooding = morphClean(rawFlooding, width, height);
  const cleanDeforestation = morphClean(rawDeforestation, width, height);
  const cleanConstruction = morphClean(rawConstruction, width, height);
  const cleanCrop = morphClean(rawCrop, width, height);

  const masks: MasksMap = {
    flooding: { width, height, data: cleanFlooding },
    deforestation: { width, height, data: cleanDeforestation },
    construction: { width, height, data: cleanConstruction },
    crop_evolution: { width, height, data: cleanCrop },
  };

  const counts: Record<string, number> = {};
  for (const [key, maskObj] of Object.entries(masks)) {
    let c = 0;
    for (let i = 0; i < maskObj.data.length; i++) {
      if (maskObj.data[i] > 0) c++;
    }
    counts[key] = (c / len) * 100;
  }

  let dominant = 'none';
  let maxPct = 0;
  const lines: string[] = [];

  for (const [key, pct] of Object.entries(counts)) {
    if (pct > 0.2) {
      lines.push(`${key.replace('_', ' ')}: ${pct.toFixed(1)}%`);
    }
    if (pct > maxPct && pct > 0.5) {
      maxPct = pct;
      dominant = key;
    }
  }

  const boxes = extractBoundingBoxes(masks, t2.width, t2.height);

  const answer =
    `Semantic change (index-driven): ` +
    (lines.length > 0 ? lines.join(', ') : 'no strong class exceeded detection thresholds') +
    `. Dominant class: ${dominant.replace('_', ' ')}.`;

  return {
    answer,
    masks,
    boxes,
    dominant,
    deltas: {
      ndvi_mean: Math.round((sumDNdvi / len) * 1000) / 1000,
      ndwi_mean: Math.round((sumDNdwi / len) * 1000) / 1000,
      ndbi_mean: Math.round((sumDNdbi / len) * 1000) / 1000,
    },
    method: 'bi-temporal NDVI/NDWI/NDBI decision tree',
    confidence: t1.crs && t2.crs ? 0.78 : 0.65,
  };
}

export function computeNdviAnomaly(
  t1: RasterScene,
  t2: RasterScene
): {
  answer: string;
  masks: MasksMap;
  boxes: BoundingBox[];
  t1Mean: number;
  t2Mean: number;
  deltaMean: number;
} {
  const width = Math.min(t1.width, t2.width);
  const height = Math.min(t1.height, t2.height);
  const len = width * height;

  const ndvi1 = computeNdvi(t1.bands.B08, t1.bands.B04);
  const ndvi2 = computeNdvi(t2.bands.B08, t2.bands.B04);

  const rawLoss = new Uint8Array(len);
  const rawGain = new Uint8Array(len);
  let sumDelta = 0;
  let sum1 = 0;
  let sum2 = 0;

  for (let i = 0; i < len; i++) {
    const d = ndvi2[i] - ndvi1[i];
    sumDelta += d;
    sum1 += ndvi1[i];
    sum2 += ndvi2[i];

    rawLoss[i] = d < -0.1 ? 1 : 0;
    rawGain[i] = d > 0.1 ? 1 : 0;
  }

  const masks: MasksMap = {
    vegetation_loss: { width, height, data: morphClean(rawLoss, width, height) },
    vegetation_gain: { width, height, data: morphClean(rawGain, width, height) },
  };

  let lossCount = 0;
  let gainCount = 0;
  for (let i = 0; i < len; i++) {
    if (masks.vegetation_loss.data[i] > 0) lossCount++;
    if (masks.vegetation_gain.data[i] > 0) gainCount++;
  }

  const lossPct = (lossCount / len) * 100;
  const gainPct = (gainCount / len) * 100;
  const deltaMean = sumDelta / len;
  const t1Mean = sum1 / len;
  const t2Mean = sum2 / len;

  const answer = `NDVI anomaly: mean Δ=${deltaMean.toFixed(3)}. Loss ${lossPct.toFixed(1)}%, gain ${gainPct.toFixed(1)}%.`;

  return {
    answer,
    masks,
    boxes: extractBoundingBoxes(masks, t2.width, t2.height),
    t1Mean: Math.round(t1Mean * 1000) / 1000,
    t2Mean: Math.round(t2Mean * 1000) / 1000,
    deltaMean: Math.round(deltaMean * 1000) / 1000,
  };
}

File: satquery/change.py
"""Semantic bi-temporal change: flood, construction, deforestation, crop."""

from __future__ import annotations

import cv2
import numpy as np
from PIL import Image

from satquery.indices import compute_index
from satquery.raster import RasterScene
from satquery.segmentation import overlay_masks, coverage_table, masks_to_boxes


CHANGE_CLASSES = ("flooding", "deforestation", "construction", "crop_evolution")


def _resize_like(arr: np.ndarray, shape: tuple[int, int]) -> np.ndarray:
    if arr.shape[:2] == shape:
        return arr
    return cv2.resize(arr.astype(np.float32), (shape[1], shape[0]), interpolation=cv2.INTER_LINEAR)


def semantic_change(t1: RasterScene, t2: RasterScene) -> dict:
    ndvi1 = compute_index(t1.bands, "ndvi").array
    ndvi2 = compute_index(t2.bands, "ndvi").array
    ndwi1 = compute_index(t1.bands, "ndwi").array
    ndwi2 = compute_index(t2.bands, "ndwi").array
    ndbi1 = compute_index(t1.bands, "ndbi").array
    ndbi2 = compute_index(t2.bands, "ndbi").array

    ndvi2 = _resize_like(ndvi2, ndvi1.shape)
    ndwi2 = _resize_like(ndwi2, ndwi1.shape)
    ndbi2 = _resize_like(ndbi2, ndbi1.shape)

    d_ndvi = ndvi2 - ndvi1
    d_ndwi = ndwi2 - ndwi1
    d_ndbi = ndbi2 - ndbi1

    flooding = ((ndwi2 > 0.05) & (d_ndwi > 0.12) & (ndvi2 < ndvi1)).astype(np.uint8)
    deforestation = ((d_ndvi < -0.15) & (ndvi1 > 0.25) & (ndwi2 < 0.1)).astype(np.uint8)
    construction = ((d_ndbi > 0.12) & (d_ndvi < -0.05) & (ndwi2 < 0.1)).astype(np.uint8)
    crop = ((np.abs(d_ndvi) > 0.12) & (ndvi1 > 0.15) & (ndvi2 > 0.05) & (ndwi2 < 0.15) & (construction == 0) & (deforestation == 0)).astype(np.uint8)

    kernel = np.ones((5, 5), np.uint8)
    masks = {
        "flooding": cv2.morphologyEx(flooding, cv2.MORPH_OPEN, kernel),
        "deforestation": cv2.morphologyEx(deforestation, cv2.MORPH_OPEN, kernel),
        "construction": cv2.morphologyEx(construction, cv2.MORPH_OPEN, kernel),
        "crop_evolution": cv2.morphologyEx(crop, cv2.MORPH_OPEN, kernel),
    }
    coverage = coverage_table(masks)
    visual = overlay_masks(t2.rgb, masks, opacity=0.5)
    dominant = max(coverage, key=coverage.get) if coverage else "none"
    lines = [f"{k.replace('_', ' ')}: {v:.1f}%" for k, v in coverage.items() if v > 0.05]
    answer = (
        "Semantic change (index-driven): "
        + (", ".join(lines) if lines else "no strong class exceeded detection thresholds")
        + f". Dominant class: {dominant.replace('_', ' ')}."
    )
    return {
        "answer": answer,
        "coverage": coverage,
        "masks": masks,
        "visual": visual,
        "boxes": masks_to_boxes(masks, t2.rgb.size),
        "dominant": dominant,
        "deltas": {
            "ndvi_mean": float(np.nanmean(d_ndvi)),
            "ndwi_mean": float(np.nanmean(d_ndwi)),
            "ndbi_mean": float(np.nanmean(d_ndbi)),
        },
        "method": "bi-temporal NDVI/NDWI/NDBI decision tree",
        "confidence": 0.68 if t1.georeferenced and t2.georeferenced else 0.52,
    }


def ndvi_anomaly(t1: RasterScene, t2: RasterScene) -> dict:
    a = compute_index(t1.bands, "ndvi")
    b = compute_index(t2.bands, "ndvi")
    b_arr = _resize_like(b.array, a.array.shape)
    delta = b_arr - a.array
    stress = (delta < -0.1).astype(np.uint8)
    gain = (delta > 0.1).astype(np.uint8)
    masks = {"vegetation_loss": stress, "vegetation_gain": gain}
    visual = overlay_masks(t2.rgb, masks)
    return {
        "answer": (
            f"NDVI anomaly: mean Δ={float(np.nanmean(delta)):.3f}. "
            f"Loss {coverage_table(masks)['vegetation_loss']:.1f}%, "
            f"gain {coverage_table(masks)['vegetation_gain']:.1f}%."
        ),
        "coverage": coverage_table(masks),
        "masks": masks,
        "visual": visual,
        "boxes": masks_to_boxes(masks, t2.rgb.size),
        "t1_ndvi_mean": a.mean,
        "t2_ndvi_mean": b.mean,
        "method": "NDVI anomaly tracking",
        "confidence": 0.7,
        "index_visual": b.visual,
    }
